src/dags/dag.py

```diff
@@ -45,19 +45,94 @@ def concatenate_functions(
         function: A function that produces targets when called with suitable arguments.
 
     """
-    _functions = _harmonize_functions(functions)
-    _targets = _harmonize_targets(targets, list(_functions))
-    _fail_if_targets_have_wrong_types(_targets)
-    _fail_if_functions_are_missing(_functions, _targets)
 
+    # Create the DAG.
+    dag = create_dag(functions, targets)
+
+    # Build combined function.
+    out = create_combined_function_from_dag(
+        dag, functions, targets, return_type, aggregator, enforce_signature
+    )
+
+    return out
+
+
+def create_dag(functions, targets):
+    """Build a directed acyclic graph (DAG) from functions.
+
+    Functions can depend on the output of other functions as inputs, as long as the
+    dependencies can be described by a directed acyclic graph (DAG).
+
+    Functions that are not required to produce the targets will simply be ignored.
+
+    Args:
+        functions (dict or list): Dict or list of functions. If a list, the function
+            name is inferred from the __name__ attribute of the entries. If a dict,
+            the name of the function is set to the dictionary key.
+        targets (str): Name of the function that produces the target or list of such
+            function names.
+
+    Returns:
+        dag: the DAG (as networkx.DiGraph object)
+
+    """
+    # Harmonize and check arguments.
+    _functions, _targets = harmonize_and_check_functions_targets(functions, targets)
+
+    # Create the DAG
     _raw_dag = _create_complete_dag(_functions)
-    _dag = _limit_dag_to_targets_and_their_ancestors(_raw_dag, _targets)
-    _arglist = _create_arguments_of_concatenated_function(_functions, _dag)
-    _exec_info = _create_execution_info(_functions, _dag)
+    dag = _limit_dag_to_targets_and_their_ancestors(_raw_dag, _targets)
+
+    # Check if there are cycles in the DAG
+    _fail_if_dag_contains_cycle(dag)
+
+    return dag
+
+
+def create_combined_function_from_dag(
```

Comment:
I think this should be a private function, because it achieves the same as `concatenate_functions` but has one more argument. Creating the dag is cheap enough to do it twice if necessary.